### PR TITLE
Add the checkers.

### PR DESCRIPTION
**Pull request by fabricematrat**

### CODE REVIEW

File: macaroonbakery/auth_context.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __hash__(self):
+        if self._hash is None:
+            self._hash = hash(frozenset(self._dict.items()))
+        return self._hash
+
+    def __eq__(self, other):
```
Comment (rogpeppe) [inline macaroonbakery/auth_context.py]: I'm not sure we need to support equality on this type.

File: macaroonbakery/checkers/caveat.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    for op in ops:
+        if op.find(' ') != -1:
+            return error_caveat('invalid operation name "{}"'.format(op))
+    return _first_party(cond, ' '.join(ops))
+
+
+def time_before_caveat(t):
+    '''Return a caveat that specifies that the time that it is checked at
+    should be before t.
+    :param t is a a UTC date in - use datetime.utcnow, not datetime.now
```
Comment (rogpeppe) [inline macaroonbakery/checkers/caveat.py]: Does this work correctly if someone *does* pass a time with a time zone in?
If not, perhaps we need to check whether there's a time zone before adding the "Z" suffix?

File: macaroonbakery/checkers/caveat.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        i = cav.index(' ')
+    except ValueError:
+        return cav, ''
+    if i == 0:
+        raise ValueError('caveat starts with space character')
+    return cav[0:i], cav[i + 1:]
+
+
+def _first_party(name, arg):
+    condition = name + ' ' + arg
```
Comment (rogpeppe) [inline macaroonbakery/checkers/caveat.py]: perhaps avoid the unnecessary concatenation when arg is empty?

    condtion = name
    if arg  != '':
        condition += ' ' + arg

?
Also, I suspect you might want to export a "condition" function like the Go package does.
I found it useful in at least one place elsewhere.

File: macaroonbakery/checkers/checkers.py [hunk 4]
```diff
@@ -1,53 +1,231 @@
...
+        if ns is None:
+            ns = Namespace()
+        self._ns = ns
+        self._checkers = {}
+        if include_std_checkers:
+            self.register_std()
+
+    def check_first_party_caveat(self, ctx, cav):
+        ''' Checks the caveat against all registered caveats conditions.
+        :return: error message if any or None
```
Comment (rogpeppe) [inline macaroonbakery/checkers/checkers.py]: s/message/message string/
?

File: macaroonbakery/checkers/checkers.py [hunk 5]
```diff
@@ -1,53 +1,231 @@
...
+        ''' Returns information on all the registered checkers.
+
+        Sorted by namespace and then name
+        :returns a list of CheckerInfo
+        '''
+        return sorted(self._checkers.values(), key=lambda x: (x._ns, x._name))
+
+    def register(self, cond, uri, check):
+        ''' Registers the given condition in the given namespace URI
+        to be checked with the given check function.
```
Comment (rogpeppe) [inline macaroonbakery/checkers/checkers.py]: Please document the expected types of the arguments and the contract of the check function.

File: macaroonbakery/checkers/declared.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
+# Copyright 2017 Canonical Ltd.
+# Licensed under the LGPLv3, see LICENCE file for details.
+
+from macaroonbakery.checkers.namespace import Namespace
+from macaroonbakery.checkers.caveat import parse_caveat
+from macaroonbakery.checkers import COND_DECLARED, STD_NAMESPACE
+
+DECLARED_KEY = 'declared-key'
```
Comment (rogpeppe) [inline macaroonbakery/checkers/declared.py]: I don't think this should be public.

File: macaroonbakery/auth_context.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        :return:
+        '''
+        new_dict = dict(self._dict)
+        new_dict[key] = val
+        return AuthContext(new_dict)
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __len__(self):
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/auth_context.py]: why not add this capability, I think it's nicer.

File: macaroonbakery/checkers/checkers.py [hunk 8]
```diff
@@ -1,53 +1,231 @@
 # Copyright 2017 Canonical Ltd.
 # Licensed under the LGPLv3, see LICENCE file for details.
 
-import collections
-import datetime
+from datetime import datetime
 
-import rfc3339
+from pyrfc3339 import parse
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/checkers/checkers.py]: done.

**Comment by fabricematrat** [comment]
:shipit:

**Pull request merged**